REGRESSION (iOS 18): Loading is blocked when trying to load localhost content from HTTPS website

basuke opened the pull request:
#### 93b599ee4b957ae9f803c133528d4bc33c88b02e
<pre>
REGRESSION (iOS 18): Loading is blocked when trying to load localhost content from HTTPS website
<a href="https://bugs.webkit.org/show_bug.cgi?id=279249">https://bugs.webkit.org/show_bug.cgi?id=279249</a>
<a href="https://rdar.apple.com/135479521">rdar://135479521</a>

Reviewed by Matthew Finkel.

Those are treated as potentially trustworthy origin. It is defined in secure context.
<a href="https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy">https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy</a>

* LayoutTests/http/tests/security/mixedContent/insecure-localhost-image-in-main-frame-UpgradeMixedContent-expected.txt:
* LayoutTests/http/tests/security/mixedContent/resources/frame-with-insecure-localhost-image.html:
* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::logConsoleWarningForUpgrade):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
(WebCore::shouldBlockInsecureContent):
* Source/WebCore/loader/MixedContentChecker.h:
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp:
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):

Canonical link: <a href="https://commits.webkit.org/291988@main">https://commits.webkit.org/291988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c417c90382b92fee8229042f00140ac05cbd22e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94446 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14038 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3818 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99465 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44973 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96496 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14334 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22466 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72078 "Found 1 new test failure: fast/css/view-transitions-nested-transparency-layers.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29401 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97448 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10658 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85276 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52408 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10352 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2968 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44288 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80598 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3069 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101509 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21500 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15686 "Found 1 new test failure: html5lib/generated/run-entities01-data.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81078 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21749 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80453 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20100 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24998 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2378 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14725 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21477 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26631 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21161 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24621 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22902 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/loader/MixedContentChecker.cpp:
```diff
@@ -206,7 +206,7 @@ bool MixedContentChecker::canModifyRequest(const URL& url, FetchOptions::Destina
         return false;
     // 4.1.2 request’s URL’s host is an IP address.
-    if (!shouldUpgradeIPAddressAndLocalhostForTesting && URL::hostIsIPAddress(url.host()))
+    if (!shouldUpgradeIPAddressAndLocalhostForTesting && !shouldTreatAsPotentiallyTrustworthy(url) && URL::hostIsIPAddress(url.host()))
```
Doesn't this make `shouldUpgradeIPAddressAndLocalhostForTesting` kind of meaningless? Since whether it's true or false, `shouldTreatAsPotentiallyTrustworthy` will return true for localhost and IP addresses.

basuke replied:
Cool. Thanks. And I think this order is much easier to read.
```
    // 4.1.2 request’s URL’s host is an IP address.
    if (URL::hostIsIPAddress(url.host()) && !shouldTreatAsPotentiallyTrustworthy(url))
```

annevk reviewed (commented):
I'd be more comfortable if @sysrqb reviewed this, but I do have a question (inline).

sysrqb reviewed (approved):
Looks okay to me, thanks

basuke committed:
REGRESSION (iOS 18): Loading is blocked when trying to load localhost content from HTTPS website
https://bugs.webkit.org/show_bug.cgi?id=279249
rdar://135479521

Reviewed by Matthew Finkel.

Those are treated as potentially trustworthy origin. It is defined in secure context.
https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy

* LayoutTests/http/tests/security/mixedContent/insecure-localhost-image-in-main-frame-UpgradeMixedContent-expected.txt:
* LayoutTests/http/tests/security/mixedContent/resources/frame-with-insecure-localhost-image.html:
* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::logConsoleWarningForUpgrade):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
(WebCore::shouldBlockInsecureContent):
* Source/WebCore/loader/MixedContentChecker.h:
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp:
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):

Canonical link: https://commits.webkit.org/291988@main
